core/src/stores/postgres.rs

```diff
@@ -2887,6 +2894,116 @@ impl Store for PostgresStore {
         Ok(())
     }
 
+    async fn upsert_data_source_table_csv(
+        &self,
+        project: &Project,
+        data_source_id: &str,
+        table_id: &str,
+        bucket: &str,
+        schema: &TableSchema,
+        rows: &Vec<Row>,
+    ) -> Result<()> {
+        let project_id = project.project_id();
+        let data_source_id = data_source_id.to_string();
+        let table_id = table_id.to_string();
+
+        let pool = self.pool.clone();
+        let c = pool.get().await?;
+
+        // Get the data source row id.
+        let stmt = c
+            .prepare(
+                "SELECT id FROM data_sources WHERE project = $1 AND data_source_id = $2 LIMIT 1",
+            )
+            .await?;
+        let r = c.query(&stmt, &[&project_id, &data_source_id]).await?;
+        let data_source_row_id: i64 = match r.len() {
+            0 => Err(anyhow!("Unknown DataSource: {}", data_source_id))?,
+            1 => r[0].get(0),
+            _ => unreachable!(),
+        };
+
+        let now = utils::now();
+
+        let field_names = schema
+            .columns()
+            .iter()
+            .map(|c| c.name.clone())
+            .collect::<Vec<_>>();
+
+        // Read all rows and upload to GCS
+        let mut wtr = Writer::from_writer(vec![]);
+        for row in rows {
+            wtr.write_record(row.to_csv_record(&field_names)?.as_slice())?;
+        }
+        let content_duration = utils::now() - now;
+        let now = utils::now();
+
+        let csv = wtr.into_inner()?;
+        let bucket_path = format!("{}.csv", table_id);
+
+        Object::create(bucket, csv, &bucket_path, "text/csv").await?;
+
+        // Update the csv bucket and path.
+        let stmt = c
+            .prepare(
+                "UPDATE tables SET csv_bucket = $1, csv_bucket_path = $2 WHERE data_source = $3 AND table_id = $4",
+            )
+            .await?;
+        c.query(
+            &stmt,
+            &[&bucket, &bucket_path, &data_source_row_id, &table_id],
+        )
+        .await?;
+
+        let upload_duration = utils::now() - now;
+
+        info!(
+            content_duration = content_duration,
+            upload_duration = upload_duration,
+            "CSV upload"
+        );
+
+        Ok(())
+    }
+
+    async fn clear_data_source_table_csv(
```

Comment:
Then this function would solely consist in deleting the object?

Reply:
Right now this method is called when we do a non-truncate upsert. So I think it's just a transitional thing. Eventually, we'll have logic that reads the file, updates it and saves it back. So it's more of a different flavor of `upsert_data_source_table_csv` than something that looks like a deletion.